lib/HTFeed/StorageZipAudit.pm

```diff
@@ -0,0 +1,199 @@
+  my $self = shift;
+
+  my ($namespace, $objid, $path, $version);
+  my $sql = 'SELECT namespace,id,path,version FROM feed_backups' .
+            ' WHERE storage_name=?' .
+            ' AND deleted IS NULL' .
+            ' AND saved_md5sum IS NOT NULL' .
+            ' AND restore_request IS NULL' .
+            ' ORDER BY RAND() LIMIT 1';
+  eval {
+    my $ref = get_dbh->selectall_arrayref($sql, undef, $self->{storage_name});
+    ($namespace, $objid, $path, $version) = @{$ref->[0]} if scalar @$ref;
+  };
+  if ($@) {
+    die "Database query failed: $@";
+  }
+  return unless $namespace && $objid && $path && $version;
+  return $self->storage_object($namespace, $objid, $path, $version);
+}
+
+
+# Get the object(s) selected for audit by choose_random_object() or some other criteria.
+# In the case of AWS Glacier, it is one or more objects selected in a previous run
+# that have been restored in the intervening time.
+# These are the objects on the filesystem ready for checksumming.
+sub all_objects {
+  my $self = shift;
+
+  return [$self->random_object()];
+}
+
+sub storage_object {
+  my $self      = shift;
+  my $namespace = shift;
+  my $objid     = shift;
+  my $path      = shift;
+  my $version   = shift;
+  
+  unless ($namespace && $objid && $path && $version) {
+    croak "invalid storage_object() args: namespace, objid, path, version required";
+  }
+  my $obj = {namespace => $namespace, objid => $objid, version => $version,
+             path => $path, storage_name => $self->{storage_name}};
+  $obj->{volume} = HTFeed::Volume->new(packagetype => 'pkgtype', namespace => $namespace, objid => $objid);
+  my $config = get_config('storage_classes');
+  $obj->{storage_config} = $config->{$self->{storage_name}};
+  $obj->{storage} = $obj->{storage_config}->{class}->new(volume => $obj->{volume},
+                                                         config => $obj->{storage_config},
+                                                         name   => $self->{storage_name});
+  $obj->{storage}->{zip_suffix} = '.gpg';
+  $obj->{storage}->{timestamp} = $obj->{version};
+  $obj->{zip_file} = $obj->{storage}->zip_filename();
+  $obj->{mets_file} = $obj->{storage}->mets_filename();
+  $obj->{zip_path} = $self->storage_object_path($obj) . '/' . $obj->{zip_file};
+  $obj->{mets_path} = $self->storage_object_path($obj) . '/' . $obj->{mets_file};
+  return $obj;
+}
+
+sub storage_object_path {
+  my $self = shift;
+  my $obj  = shift;
+
+  return $obj->{path};
+}
+
+sub run {
+  my $self = shift;
+  my $obj  = shift;
+
+  my $error_count = 0;
+  my $objects = (defined $obj) ? [$obj] : $self->all_objects();
+  unless (0 < scalar @$objects) {
+    get_logger->trace("no auditable files: returning");
+    return $error_count;
+  }
+  foreach my $obj (@$objects) {
+    eval {
+      die "encrypted zip $obj->{zip_path} does not exist" unless -f $obj->{zip_path};
+      my $db_ok = $self->check_encrypted_zip_against_database($obj);
```

Comment:
I think it's fine for now to assume the zip is encrypted, but we wouldn't be able to make that assumption if we use this for auditing the primary repository as well

Reply:
I'm replacing the call to storage's `validate_zip_checksum` with `validate_zip` so it will be encryption agnostic going forward.